File: My Astar/Main_A_Star.m
clc; clear;
xrange=100; yrange=100;    %grid coarse or fine by increasing/decreasing number of elements on map
axis([1 xrange 1 yrange]);
grid on;
hold on;
% 1-space(unvisited), 0-obstacles, 2-space(visited), 3-target, 4-start 
map=ones(xrange,yrange);
n_obstacles=randi(((xrange)^2)/2,1);  %number of obstacles
obs_x=randi(xrange,n_obstacles,1); obs_y=randi(xrange,n_obstacles,1);  %random positions of obstables
obs=[obs_x obs_y];  %obstacles coordinates
% insert obstacles in map
for i=1:n_obstacles
    map(obs_x(i),obs_y(i))=0;
end
endpoints=randi(xrange,2,2);    %target and start coordinates
target=endpoints(1,:);
map(endpoints(1,1),endpoints(1,2))=3;    %mark target coordinates on map
plot(obs_x,obs_y,'or');
start=endpoints(2,:);
map(endpoints(2,1),endpoints(2,2))=4; 

current=start;         %defining start node as the current node
path_distance = dist_calc(start(1),start(2),current(1),current(2));   %current hn
path = current;                                                       %all nodes of the path 
successor = neighbors(start(1),start(2),xrange,map);     
visited=succ_dist(current(1),current(2),successor,map,path_distance);

[node,node_dist]=minimum_distance(current(1),current(2),path_distance,successor,map,visited);

path_distance=path_distance+node_dist;
path=[path;node];

current=node;
while map(current(1),current(2))~=3
successor = neighbors(current(1),current(2),xrange,map);
visit_temp=succ_dist(current(1),current(2),successor,map,path_distance);
visited=[visited;visit_temp];
[node,node_dist]=minimum_distance(current(1),current(2),path_distance,successor,map,visited);
path_distance=path_distance+node_dist;
path=[path;node];
current=node;
end

plot(path(:,1),path(:,2),'*k');

File: My Astar/Python/Main_A_Star.py
import numpy as np
from math import *
import matplotlib.pyplot as plt 
from astarfunctions import *


xrange=100; yrange=100;  #x and y axis limits

# 1-space(unvisited), 0-obstacles, 2-space(visited), 3-target, 4-start 

map=np.ones((xrange,yrange));  #introducing map matrix

n_obstacles=np.random.randint((((xrange)**2)/2));  #number of obstacles
n_obstacles=n_obstacles + 1;   #to avoid zero
obs_x=np.random.choice(np.arange(xrange),size=(n_obstacles));
obs_y=np.random.choice(np.arange(yrange),size=(n_obstacles));
obs=np.vstack([obs_x,obs_y]).T;   #co-ordinates of the obstacles
plt.plot([obs_x],[obs_y],'o',color='r');

for i in range(n_obstacles):
    map[obs_x[i],obs_y[i]] = 0;
    
endpoints=np.random.randint(xrange,size=(2,2));
target=endpoints[0,:];
map[target[0],target[1]]=3;
start=endpoints[1,:];
map[start[0],start[1]]=4;
plt.plot(target[0],target[1],'o');
plt.plot(start[0],start[1],'o');

current=start;

path_distance=dist_calc(start[0],start[1],current[0],current[1]);
path=current;

successor = neighbors(start[0],start[1],xrange,map);

visited=succ_dist(current[0],current[1],successor,map,path_distance);

node,node_dist=minimum_distance(current[0],current[1],path_distance,successor,map,visited);

path_distance=path_distance+node_dist;
path=np.vstack((path,node));

current=node;

counter = 0;
while map[current[0],current[1]] !=3:
    if counter>10**4:
        break;
    successor = neighbors(current[0],current[1],xrange,map);
    visit_temp=succ_dist(current[0],current[1],successor,map,path_distance);
    visited=np.vstack((visited,visit_temp));
    node,node_dist=minimum_distance(current[0],current[1],path_distance,successor,map,visited);
    path_distance=path_distance+node_dist;
    path=np.vstack((path,node));
    current=node;
    counter=counter+1;
    
plt.plot(path[:,0],path[:,1],'*');